GH-135379: Top of stack caching for the JIT.

markshannon opened the pull request:
The stats need fixing and the generated tables could be more compact, but it works.


<!-- gh-issue-number: gh-135379 -->
* Issue: gh-135379
<!-- /gh-issue-number -->

Fidget-Spinner commented on Python/optimizer.c:
```diff
@@ -1258,6 +1287,48 @@ int effective_trace_length(_PyUOpInstruction *buffer, int length)
+        }
+        assert(_PyUop_Caching[uop].opcodes[depth] != 0);
+        assert(spill_or_reload->opcode != 0);
+        buffer[i*2+1].opcode = _PyUop_Caching[uop].opcodes[depth];
```
It would be good to add custom cases to better handle COPY and SWAP later via replication. But we can do that in another PR.

markshannon replied:
done

Fidget-Spinner reviewed (commented):
This is really cool. I'll do a full review soon enough.

Fidget-Spinner commented on Misc/NEWS.d/next/Core_and_Builtins/2025-06-20-16-03-59.gh-issue-135379.eDg89T.rst:
```diff
@@ -0,0 +1,6 @@
+in registers. The code generator generates multiple versions of thos uops
+that do not escape and are relatively small. During JIT compilation, the
+copy that produces the least memory traffic is selected, spilling or
+reloading values when needed.
```
```suggestion
Implement a limited form of register allocation known as "top of stack
caching" in the JIT. It works by keeping 0-3 of the top items in the stack
in registers. The code generator generates multiple versions of those uops
that do not escape and are relatively small. During JIT compilation, the
copy that produces the least memory traffic is selected, spilling or
reloading values when needed.
```

Fidget-Spinner commented on Misc/NEWS.d/next/Core_and_Builtins/2025-06-13-13-32-16.gh-issue-135379.pAxZgy.rst:
```diff
@@ -0,0 +1,3 @@
+Implement top-of-stack caching for the JIT (and tier 2 interpreter). Reduces
```
Why is there a second news file?

markshannon replied:
Because it's so newsworthy 😉 
I'll remove this one.

Fidget-Spinner commented on Python/optimizer.c:
```diff
@@ -995,16 +997,43 @@ count_exits(_PyUOpInstruction *buffer, int length)
 }
 
+static int
+get_exit_depth(_PyUOpInstruction *inst)
```
Can you write a short snippet on what this does? It's rather confusing otherwise. IIUC, it finds what is the number of "used" registers on exit right?

markshannon replied:
Where "exit" means a side exit (EXIT_IF or DEOPT_IF) not the end of the uop.

I've renamed it, and commented it.

Fidget-Spinner commented on Python/optimizer.c:
```diff
@@ -995,16 +997,43 @@ count_exits(_PyUOpInstruction *buffer, int length)
+        return -1;
+    }
+    if (_PyUop_Caching[base_opcode].exit_depth_is_output) {
+        return input + _PyUop_Caching[base_opcode].delta;
```
What does this do?

markshannon replied:
Returns the output stack depth, which is the input stack depth + the delta.

Fidget-Spinner commented on Python/optimizer.c:
```diff
@@ -1258,6 +1294,48 @@ int effective_trace_length(_PyUOpInstruction *buffer, int length)
+static int
+stack_allocate(_PyUOpInstruction *buffer, int length)
+{
+    for (int i = length-1; i >= 0; i--) {
```
To my understanding, this is due to the possibility of needing to insert a spill between every instruction right, so you need to reserve 2N number of instructions?

markshannon replied:
Yes.

Fidget-Spinner commented on Tools/cases_generator/analyzer.py:
```diff
@@ -1215,6 +1215,77 @@ def analyze_forest(forest: list[parser.AstNode]) -> Analysis:
+    if ideal_inputs > 3:
+        ideal_inputs = 3
+    if ideal_outputs > 3:
+        ideal_outputs = 3
```
Can you move the value `3` to a global magic number so that we can play around with increasing/decreasing register counts in the future?

Fidget-Spinner commented on Tools/cases_generator/analyzer.py:
```diff
@@ -1215,6 +1215,77 @@ def analyze_forest(forest: list[parser.AstNode]) -> Analysis:
+    #if has_exit and ideal_inputs != ideal_outputs:
+    #    n = min(ideal_inputs, ideal_outputs)
+    #    yield n, n
+    #    return
```
Let's remove this.

Fidget-Spinner reviewed (commented):
I need to review the cases generator later.

Fidget-Spinner commented on Tools/cases_generator/analyzer.py:
```diff
@@ -1215,6 +1217,71 @@ def analyze_forest(forest: list[parser.AstNode]) -> Analysis:
+        inputs, outputs = 0, outputs-inputs
+    else:
+        inputs, outputs = inputs-outputs, 0
+    while inputs <= 3 and outputs <= 3:
```
The 3 here should be replaced by `MAX_CACHED_REGISTER`?

Fidget-Spinner commented on Tools/cases_generator/tier2_generator.py:
```diff
@@ -60,15 +61,19 @@ def declare_variables(uop: Uop, out: CWriter) -> None:
 class Tier2Emitter(Emitter):
 
-    def __init__(self, out: CWriter, labels: dict[str, Label]):
+    def __init__(self, out: CWriter, labels: dict[str, Label], exit_cache_depth:int):
```
```suggestion
    def __init__(self, out: CWriter, labels: dict[str, Label], exit_cache_depth: int):
```

Fidget-Spinner commented on Tools/cases_generator/tier2_generator.py:
```diff
@@ -133,7 +141,7 @@ def oparg(
 
 
-def write_uop(uop: Uop, emitter: Emitter, stack: Stack) -> Stack:
+def write_uop(uop: Uop, emitter: Emitter, stack: Stack, cached_items:int=0) -> Stack:
```
```suggestion
def write_uop(uop: Uop, emitter: Emitter, stack: Stack, cached_items: int = 0) -> Stack:
```

Fidget-Spinner commented on Tools/cases_generator/analyzer.py:
```diff
@@ -1215,6 +1217,71 @@ def analyze_forest(forest: list[parser.AstNode]) -> Analysis:
+def is_large(uop: Uop) -> bool:
+    return len(list(uop.body.tokens())) > 80
+
+def get_uop_cache_depths(uop: Uop) -> Iterator[tuple[int, int, int]]:
```
I don't like using generators here. There's no benefit as it's not like the uop sequence is huge or something. The code might be clearer without the yields and with just a list append instead. This is an optional review though.

markshannon replied:
I think using a generator here is fine.

Fidget-Spinner reviewed (commented):
Stylistic nits

Fidget-Spinner reviewed (approved):
I was discussing with Brandt the other day, and I concluded we need this to implement decref elimination with lower maintenance burden.

For decref elimination, there are two ways:
1. Manually handwrite/cases generate a version of the instruction that has no PyStackref_CLOSE/DECREF_INPUTS.
2. Specialize for POP_TOP.

Option 2 is a lot more maintainable in the long run and doesn't involve any more cases generator magic like in 1. It's also less likely to introduce a bug, as again less cases generator magic. So I'm more inclined to 2.

Option 2. requires TOS caching, otherwise it won't pay off. So this PR is needed otherwise we're blocking other optimizations. Unless of course, you folks don't mind me exercising some cases generator magic again :wink: .

efimov-mikhail commented on Tools/cases_generator/analyzer.py:
```diff
@@ -1236,6 +1303,7 @@ def dump_analysis(analysis: Analysis) -> None:
         p.dump("    ")
 
 
+
```
```suggestion
```

diegorusso commented on Include/internal/pycore_optimizer.h:
```diff
@@ -119,7 +119,7 @@ PyAPI_FUNC(void) _Py_Executors_InvalidateCold(PyInterpreterState *interp);
 
 // This is the length of the trace we project initially.
-#define UOP_MAX_TRACE_LENGTH 800
+#define UOP_MAX_TRACE_LENGTH 1600
```
Silly question, why are we doubling this up?

Fidget-Spinner replied:
Every instruction might now require a spill instruction before it. So for n instructions we now have 2n ones at maximum.

diegorusso commented on Misc/NEWS.d/next/Core_and_Builtins/2025-12-05-15-59-03.gh-issue-135379.lDXbKO.rst:
```diff
@@ -0,0 +1,6 @@
+Implement a limited form of register allocation known as "top of stack
+caching" in the JIT. It works by keeping 0-3 of the top items in the stack
```
> It works by keeping 0-3

Any reason why 0-3? Cannot this be configurable depending on the platform?

markshannon replied:
We use set `MAX_CACHED_REGISTER = 3` in the code generator, so it shouldn't be much work to generalize.
I'd like to get this in and then look to generalize it after that.

It is possible that 3 is too high for x86 and too low for ARM. We'll need to experiment.

Fidget-Spinner commented on Python/bytecodes.c:
```diff
@@ -2324,20 +2324,21 @@ dummy_func(
                 /* Designed to work in tandem with CALL, pushes two values. */
-                *attr = PyStackRef_NULL;
-                int is_meth = _PyObject_GetMethodStackRef(tstate, PyStackRef_AsPyObjectBorrow(owner), name, attr);
+                _PyStackRef method = PyStackRef_NULL;
```
This isn't safe to do in FT builds

markshannon replied:
What isn't safe? An assignment?

Fidget-Spinner replied:
This and the line below: you can't assign a stackref to something the GC can't see.

markshannon replied:
That's a *terrible* API. Why not take a `PyObject **` if a `PyStackRef *` isn't safe?

markshannon replied:
Or, return a `PyStackRef`, so the caller doesn't need to worry about ensuring that the thing pointed to is visible to the GC?
```C
_PyStackRef _PyObject_GetMethodStackRef(PyThreadState *ts, PyObject *obj, PyObject *name, int* found);
```

Fidget-Spinner replied:
The second option sounds like it won't work for something that is as complicated as this https://github.com/python/cpython/blob/089a324a4258fd5783b7e65acc29231d1042646e/Objects/object.c#L1692

The first option doesn't work because PyObject * is refcounted, _PyStackRef can be skipped, and the refcount contention on FT is an issue.

Fidget-Spinner replied:
In the first place, I don't get why this is considered a perf critical path. LOAD_ATTR shouldn't be that common as most of the JIT stuff should be using the specialized version?

colesbury commented on Python/bytecodes.c:
```diff
@@ -2324,20 +2324,26 @@ dummy_func(
+            #ifdef Py_GIL_DISABLED
+                    method.ref = PyStackRef_NULL;
+                    _PyThreadState_PopCStackRef(tstate, &method);
+            #endif
```
Can we encapsulate this in a function? We want to unlink the `_PyCStackRef` (in the FT build), but steal the stack ref instead of closing it:

Something like:

```c
attr = _PyThreadState_StealCStackRef(tstate, &method);

...

static inline _PyStackRef
_PyThreadState_StealCStackRef(PyThreadState *tstate, _PyCStackRef *ref)
{
#ifdef Py_GIL_DISABLED
    _PyThreadStateImpl *tstate_impl = (_PyThreadStateImpl *)tstate;
    assert(tstate_impl->c_stack_refs == ref);
    tstate_impl->c_stack_refs = ref->next;
#endif
    return ref->ref;
}
```

With whatever naming you think is appropriate.

colesbury reviewed (commented):
One comment below. The `_LOAD_ATTR` changes look fine to me.

savannahostrowski commented on Include/internal/pycore_optimizer.h:
```diff
@@ -36,7 +36,7 @@ typedef struct {
     uint8_t chain_depth:6;  // Must be big enough for MAX_CHAIN_DEPTH - 1.
     bool warm;
-    int index;           // Index of ENTER_EXECUTOR (if code isn't NULL, below).
+    int16_t index;           // Index of ENTER_EXECUTOR (if code isn't NULL, below).
```
Was this intentional? In practice, functions are very small and should probably never exceed this, but do we want a check here or some validation? I think this could overflow silently, right?

markshannon replied:
Restored to `int32_t`

savannahostrowski reviewed (approved):
Took a quick skim of this - very neat! Thanks for also including all the perf numbers in the discussion, which helps counteract some of the initial trepidation I had around the amount of new generated code. This lays pretty solid groundwork for future optimizations as well.

Just one comment about the type change in `pycore_optimizer.h`

Fidget-Spinner commented on Tools/cases_generator/analyzer.py:
```diff
@@ -1280,6 +1288,70 @@ def analyze_forest(forest: list[parser.AstNode]) -> Analysis:
 
+#Simple heuristic for size to avoid too much stencil duplication
+def is_large(uop: Uop) -> bool:
+    return len(list(uop.body.tokens())) > 80
```
This is too low, to be useful. For example, `_BINARY_OP_ADD_FLOAT` is 89 tokens. It should be register allocated, but instead it spills all the time because it's limited to `r2_1`. I've inspected the code for nbody and it spills unecessarily. We should up the limit.

I suggest increasing the limit to 150.

Fidget-Spinner replied:
Some benchmark results for nbody tracing jit vs nbody tracing jit + regalloc after increasing the limit. Base commits are a few apart but they have no JIT related changes in them:
```
Mean +- std dev: [nbody_tracing] 107 ms +- 2 ms -> [nbody_tracing_regalloc] 78.8 ms +- 0.6 ms: 1.36x faster
```
36% faster. That's significantly higher than the 17-27% we've been getting.

markshannon replied:
Changed to 150.

markshannon committed:
Tier 2 TOS caching

Uses three registers to cache values at the top of the evaluation stack
This significantly reduces memory traffic for smaller, more common uops.

markshannon committed:
Merge branch 'main' into tier-2-tos-caching

Fidget-Spinner reviewed (approved):
I've read the code and understood it mostly as I had to rebase it a few times on my own branch.

These are the latest benchmarking results for this branch https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20251210-3.15.0a2%2B-4b24c15-JIT/bm-20251210-vultr-x86_64-faster%252dcpython-tier_2_tos_caching-3.15.0a2%2B-4b24c15-vs-base.md

The 19% nbody speedup is still there, along with a modest ~7% speedup for richards as well.

So this bodes very well for the JIT. Considering it will unlock future optimizations in decref elimination, which would be very hard without this, we should just go ahead and merge it.

markshannon committed:
Merge branch 'main' into tier-2-tos-caching
